// openclaw/dist/plugin-sdk/browser-node-runtime.js
import { n as defaultRuntime } from "../runtime-izpjJukX.js";
import { i as runExec } from "../exec-C4nQnfsK.js";
import { a as isLoopbackHost } from "../net-BBCaEpfz.js";
import { n as resolveGatewayAuth } from "../auth-resolve-D9vCjzKT.js";
import { t as rawDataToString } from "../ws-CQQhqVXc.js";
import { Mr as errorShape, jr as ErrorCodes } from "../protocol-Hjar_s3V.js";
import { t as withTimeout } from "../with-timeout-JkpU6HA2.js";
import { n as startLazyPluginServiceModule } from "../lazy-service-module-2HZrBZvK.js";
import { n as callGatewayFromCli, t as addGatewayClientOptions } from "../gateway-rpc-DGmyKiL0.js";
import { n as runCommandWithRuntime } from "../cli-utils-Bey8_EmD.js";
import { i as resolveNodeCommandAllowlist, n as isNodeCommandAllowed } from "../node-command-policy-DrAl5pZQ.js";
import { t as safeParseJson } from "../server-json-DCqyamk1.js";
import { n as respondUnavailableOnNodeInvokeError } from "../nodes.helpers-BvSQb3Px.js";
import { t as ensureGatewayStartupAuth } from "../startup-auth-C0ghpvlY.js";
import "../browser-node-runtime-ChFzaJ-a.js";
export { ErrorCodes, addGatewayClientOptions, callGatewayFromCli, defaultRuntime, ensureGatewayStartupAuth, errorShape, isLoopbackHost, isNodeCommandAllowed, rawDataToString, resolveGatewayAuth, resolveNodeCommandAllowlist, respondUnavailableOnNodeInvokeError, runCommandWithRuntime, runExec, safeParseJson, startLazyPluginServiceModule, withTimeout };
