// openclaw/dist/extensions/github-copilot/register.runtime.js
import { a as coerceSecretRef } from "../../types.secrets-ClP-vJ-P.js";
import { n as ensureAuthProfileStore } from "../../store-D-8DaAtv.js";
import { n as listProfilesForProvider } from "../../profile-list-zV5Cv5VC.js";
import "../../provider-auth-LNc11avL.js";
import { r as resolveCopilotApiToken, t as DEFAULT_COPILOT_API_BASE_URL } from "../../github-copilot-token-C5l48y20.js";
import { t as githubCopilotLoginCommand } from "../../provider-auth-login-DfqBjAYS.js";
import { n as resolveCopilotForwardCompatModel, t as PROVIDER_ID } from "../../models-CVzS8osi.js";
import { r as wrapCopilotProviderStream, t as wrapCopilotAnthropicStream } from "../../stream-BrNQazbE.js";
import "../../token-C2FDsyfj.js";
import { t as fetchCopilotUsage } from "../../usage-go9wKgOC.js";
export { DEFAULT_COPILOT_API_BASE_URL, PROVIDER_ID, coerceSecretRef, ensureAuthProfileStore, fetchCopilotUsage, githubCopilotLoginCommand, listProfilesForProvider, resolveCopilotApiToken, resolveCopilotForwardCompatModel, wrapCopilotAnthropicStream, wrapCopilotProviderStream };
